servant-server/src/Servant/Server/Internal.hs

```diff
@@ -134,8 +135,8 @@ processMethodRouter :: forall a. ConvertibleStrings a B.ByteString
                     -> Maybe [(HeaderName, B.ByteString)]
                     -> Request -> RouteResult Response
 processMethodRouter handleA status method headers request = case handleA of
-  Nothing -> failWith UnsupportedMediaType
-  Just (contentT, body) -> succeedWith $ responseLBS status hdrs bdy
+  Nothing -> FailFatal err406
```

Comment:
I'm going to keep arguing just because it helps to clarify things. Ok, so let's say instead of:

> type MyAPI c = ... Put '[c] T ...

You have to write

> type MyAPI c = ... Put c T ...

Was much lost in the process?

(As a side note, we *can* rule out the pattern - we can just run some checks over the API type if we need to.)